File: Shortest Path/Dijkstra/1504번 - 특정한 최단 경로/1504.cpp
#include <iostream>
#include <vector>
#include <queue>

using namespace std;

#define MAX 801
#define INF 100000000

int n, e, v1, v2;
vector<pair<int, int> > v[MAX];
int min_val[MAX];

void input() {
    cin >> n >> e;
    for(int i = 0, a, b, c; i < e; i++) {
        cin >> a >> b >> c;
        v[a].push_back({c, b});
        v[b].push_back({c, a});
    }
    cin >> v1 >> v2;
}

void dijkstra(int st) {
    fill_n(&min_val[0], MAX, INF);
    
    priority_queue<pair<int, int> > pq;
    pq.push({0, st});
    min_val[st] = 0;
    
    while(!pq.empty()) {
        int node = pq.top().second;
        int weight = -pq.top().first;
        pq.pop();
        
        if(min_val[node] < weight) continue;
        
        for(int i = 0; i < v[node].size(); i++) {
            int next_node = v[node][i].second;
            int next_weight = v[node][i].first + weight;
            
            if(min_val[next_node] > next_weight) {
                min_val[next_node] = next_weight;
                pq.push({-next_weight, next_node});
            }
        }
    }
}

void solve() {
    int case_1 = 0, case_2 = 0;
    
    dijkstra(1);
    case_1 += min_val[v1];
    case_2 += min_val[v2];
    
    dijkstra(v1);
    case_1 += min_val[v2];
    case_2 += min_val[n];
    
    dijkstra(v2);
    case_1 += min_val[n];
    case_2 += min_val[v1];
    
    int ans = min(case_1, case_2);
    cout << (ans >= INF ? -1 : ans) << '\n';
}

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);
    
    input();
    solve();
    
    return 0;
}

File: Shortest Path/Dijkstra/1504번 - 특정한 최단 경로/1504.py
import sys
import heapq

input = sys.stdin.readline

n, e = map(int, input().split())
arr = [list() for _ in range(n+1)]
for _ in range(e):
    a, b, c = map(int, input().split())
    arr[a].append([b, c])
    arr[b].append([a, c])
v1, v2 = map(int, input().split())
INF = 10**7

def dijkstra(st):
    min_val = [INF] * (n + 1)
    min_val[st] = 0
    hq = []
    heapq.heappush(hq, [0, st])
    
    while hq:
        weight, node = heapq.heappop(hq)
        
        if min_val[node] < weight:
            continue
        
        for info in arr[node]:
            next_node = info[0]
            next_weight = info[1] + weight
            
            if min_val[next_node] > next_weight:
                min_val[next_node] = next_weight
                heapq.heappush(hq, [next_weight, next_node])
    
    return min_val

def solve():
    case_1, case_2 = 0, 0
    
    dij_1 = dijkstra(1)
    case_1 += dij_1[v1]
    case_2 += dij_1[v2]
    
    dij_v1 = dijkstra(v1)
    case_1 += dij_v1[v2]
    case_2 += dij_v1[n]
    
    dij_v2 = dijkstra(v2)
    case_1 += dij_v2[n]
    case_2 += dij_v2[v1]
    
    ans = min(case_1, case_2)
    print(ans if ans < INF else -1)

if __name__ == '__main__':
    solve()